### ISSUE TITLE
[JENKINS-45771] Disable shallow clone when we know a merge will take place

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3160]
<p>There is an issue when building pull requests and the clone trait is present with shallow clone option.<br/>
 Apparently you need some history in order to be able to do a merge.</p>

<p>We need to disable shallow clone when we know a merge will take place.</p>

<p>At the first look, this can be fixed in<br/>
 <tt>com.cloudbees.jenkins.plugins.bitbucket.BitbucketGitSCMBuilder</tt><br/>
 <tt>org.jenkinsci.plugins.github_branch_source.GitHubSCMBuilder</tt></p>

<p>I had identified the problem using bitbucket-branch-source-plugin but the should be visible on github-branch-source-plugin also..</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drealecs.png" title="drealecs's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=drealecs">drealecs</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771" target="_blank">Disable shallow clone when we know a merge will take place</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nfalco">nfalco</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: bitbucket-branch-source-plugin, git-plugin, github-branch-source-plugin
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2024-12-05T02:30:31+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>There is an issue when building pull requests and the clone trait is present with shallow clone option.
 Apparently you need some history in order to be able to do a merge.</p>

<p>We need to disable shallow clone when we know a merge will take place.</p>

<p>At the first look, this can be fixed in
 <tt>com.cloudbees.jenkins.plugins.bitbucket.BitbucketGitSCMBuilder</tt>
 <tt>org.jenkinsci.plugins.github_branch_source.GitHubSCMBuilder</tt></p>

<p>I had identified the problem using bitbucket-branch-source-plugin but the should be visible on github-branch-source-plugin also..</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote links associated with this issue:_


  - [git-client-plugin#​255](https://github.com/jenkinsci/git-client-plugin/pull/255)
  - [git-plugin#​516](https://github.com/jenkinsci/git-plugin/pull/516)
  - [gitea-plugin#​4](https://github.com/jenkinsci/gitea-plugin/pull/4)
  - [bitbucket-branch-source-plugin#​64](https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/64)
  - [github-branch-source-plugin#​150](https://github.com/jenkinsci/github-branch-source-plugin/pull/150)
  - [Page](https://wiki.jenkins.io/pages/viewpage.action?pageId=19070990)
  - [Page](https://wiki.jenkins.io/pages/viewpage.action?pageId=83361877)
  - [Page](https://wiki.jenkins.io/pages/viewpage.action?pageId=96076426)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308493&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308493">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Since the "shallowness" of a git repository is captured in the local git repository and the <tt>&#45;&#45;shallow</tt> and <tt>&#45;&#45;depth x</tt> arguments to command line git only set the configuration for future command executions, the git client internals may need to be extended to include some form of check to switch a repository from "shallow" to "unshallow".  </p>

<p>Some of the complexities of converting from shallow to unshallow are described in <a href="https://stackoverflow.com/questions/6802145/convert-shallow-clone-to-full-clone" class="external-link" target="_blank" rel="nofollow noopener">stackoverflow</a>.</p>
</pre>
</details>

</ul>
</details>

<p>Since the "shallowness" of a git repository is captured in the local git repository and the <tt>&#45;&#45;shallow</tt> and <tt>&#45;&#45;depth x</tt> arguments to command line git only set the configuration for future command executions, the git client internals may need to be extended to include some form of check to switch a repository from "shallow" to "unshallow".  </p>

<p>Some of the complexities of converting from shallow to unshallow are described in <a href="https://stackoverflow.com/questions/6802145/convert-shallow-clone-to-full-clone" class="external-link" target="_blank" rel="nofollow noopener">stackoverflow</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308494&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308494">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> my fix sets the shallowness from the start... a PR-merge will always be non-shallow and not chop and change back and forth</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> my fix sets the shallowness from the start... a PR-merge will always be non-shallow and not chop and change back and forth</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308495&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308495">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> that sounds great to me.  The general case of a user toggling between shallow and not shallow will continue to be an open issue, and the easy work around for that case is to wipe the workspace.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a> that sounds great to me.  The general case of a user toggling between shallow and not shallow will continue to be an open issue, and the easy work around for that case is to wipe the workspace.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308527&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308527">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Stephen Connolly
Path:
 pom.xml
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
 src/main/java/org/jenkinsci/plugins/gitclient/CloneCommand.java
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java
 src/main/java/org/jenkinsci/plugins/gitclient/PushCommand.java
 src/main/java/org/jenkinsci/plugins/gitclient/RevListCommand.java
<a href="http://jenkins-ci.org/commit/git-client-plugin/d2243416d262ac9b442945d05cc2cf827e5e5373" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/d2243416d262ac9b442945d05cc2cf827e5e5373</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Ensure all command boolean options can be toggled, not just set once</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 pom.xml<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/CloneCommand.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/PushCommand.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/RevListCommand.java<br/>
<a href="http://jenkins-ci.org/commit/git-client-plugin/d2243416d262ac9b442945d05cc2cf827e5e5373" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/d2243416d262ac9b442945d05cc2cf827e5e5373</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Ensure all command boolean options can be toggled, not just set once</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308528&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308528">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Stephen Connolly
Path:
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java
<a href="http://jenkins-ci.org/commit/git-client-plugin/1a7d53e90e23c6184b9b54003ef0d1b6379f8c30" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/1a7d53e90e23c6184b9b54003ef0d1b6379f8c30</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Code review catches bugs (who knew)</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java<br/>
<a href="http://jenkins-ci.org/commit/git-client-plugin/1a7d53e90e23c6184b9b54003ef0d1b6379f8c30" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/1a7d53e90e23c6184b9b54003ef0d1b6379f8c30</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Code review catches bugs (who knew)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308529&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308529">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Stephen Connolly
Path:
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
 src/main/java/org/jenkinsci/plugins/gitclient/FetchCommand.java
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java
<a href="http://jenkins-ci.org/commit/git-client-plugin/f445161be299b3cf0d54e893cfa7df39bb0058ee" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/f445161be299b3cf0d54e893cfa7df39bb0058ee</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Missed one boolean builder</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/FetchCommand.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java<br/>
<a href="http://jenkins-ci.org/commit/git-client-plugin/f445161be299b3cf0d54e893cfa7df39bb0058ee" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/f445161be299b3cf0d54e893cfa7df39bb0058ee</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Missed one boolean builder</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308530&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308530">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Mark Waite
Path:
 pom.xml
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
 src/main/java/org/jenkinsci/plugins/gitclient/CloneCommand.java
 src/main/java/org/jenkinsci/plugins/gitclient/FetchCommand.java
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java
 src/main/java/org/jenkinsci/plugins/gitclient/PushCommand.java
 src/main/java/org/jenkinsci/plugins/gitclient/RevListCommand.java
<a href="http://jenkins-ci.org/commit/git-client-plugin/d7ed207c99feecb5d0fc854e05ff0c3df2cd275e" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/d7ed207c99feecb5d0fc854e05ff0c3df2cd275e</a>
Log:
  Merge pull request #255 from stephenc/jenkins-45771</p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Ensure all command boolean options can be toggled, not just set once</p>


<p>Compare: <a href="https://github.com/jenkinsci/git-client-plugin/compare/b0fa8071bdc1...d7ed207c99fe" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-client-plugin/compare/b0fa8071bdc1...d7ed207c99fe</a></p>

</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Mark Waite<br/>
Path:<br/>
 pom.xml<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/CloneCommand.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/FetchCommand.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/PushCommand.java<br/>
 src/main/java/org/jenkinsci/plugins/gitclient/RevListCommand.java<br/>
<a href="http://jenkins-ci.org/commit/git-client-plugin/d7ed207c99feecb5d0fc854e05ff0c3df2cd275e" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-client-plugin/d7ed207c99feecb5d0fc854e05ff0c3df2cd275e</a><br/>
Log:<br/>
  Merge pull request #255 from stephenc/jenkins-45771</p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Ensure all command boolean options can be toggled, not just set once</p>


<p>Compare: <a href="https://github.com/jenkinsci/git-client-plugin/compare/b0fa8071bdc1...d7ed207c99fe" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-client-plugin/compare/b0fa8071bdc1...d7ed207c99fe</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308572&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308572">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Stephen Connolly
Path:
 pom.xml
 src/main/java/jenkins/plugins/git/MergeWithGitSCMExtension.java
<a href="http://jenkins-ci.org/commit/git-plugin/b8c90a9461dc57633d97490d421a52796b0be2bb" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/b8c90a9461dc57633d97490d421a52796b0be2bb</a>
Log:
  <span class="error">&#91;FIXED JENKINS-45771&#93;</span> Consolidate the MergeWithGitSCMExtension into git plugin</p>

<p>With the update to git-client we can now fix the shallow clone issue on merge with</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 pom.xml<br/>
 src/main/java/jenkins/plugins/git/MergeWithGitSCMExtension.java<br/>
<a href="http://jenkins-ci.org/commit/git-plugin/b8c90a9461dc57633d97490d421a52796b0be2bb" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/git-plugin/b8c90a9461dc57633d97490d421a52796b0be2bb</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-45771&#93;</span> Consolidate the MergeWithGitSCMExtension into git plugin</p>

<p>With the update to git-client we can now fix the shallow clone issue on merge with</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308581&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308581">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Stephen Connolly
Path:
 pom.xml
 src/main/java/org/jenkinsci/plugin/gitea/GiteaSCMBuilder.java
 src/main/java/org/jenkinsci/plugin/gitea/MergeWithGitSCMExtension.java
<a href="http://jenkins-ci.org/commit/gitea-plugin/ed08e75e55f227e7d001dfd84a234fd477c7b831" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/gitea-plugin/ed08e75e55f227e7d001dfd84a234fd477c7b831</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Pick up consolidated MergeWithGitSCMExtension from git plugin</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 pom.xml<br/>
 src/main/java/org/jenkinsci/plugin/gitea/GiteaSCMBuilder.java<br/>
 src/main/java/org/jenkinsci/plugin/gitea/MergeWithGitSCMExtension.java<br/>
<a href="http://jenkins-ci.org/commit/gitea-plugin/ed08e75e55f227e7d001dfd84a234fd477c7b831" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/gitea-plugin/ed08e75e55f227e7d001dfd84a234fd477c7b831</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Pick up consolidated MergeWithGitSCMExtension from git plugin</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308584&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308584">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Stephen Connolly
Path:
 pom.xml
 src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMBuilder.java
 src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java
 src/main/java/org/jenkinsci/plugins/github_branch_source/MergeWithGitSCMExtension.java
 src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMBuilderTest.java
<a href="http://jenkins-ci.org/commit/github-branch-source-plugin/3256676fa6a47d342a0442c59555df9eecde0b68" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/github-branch-source-plugin/3256676fa6a47d342a0442c59555df9eecde0b68</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Pick up consolidated MergeWithGitSCMExtension from git plugin</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 pom.xml<br/>
 src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMBuilder.java<br/>
 src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java<br/>
 src/main/java/org/jenkinsci/plugins/github_branch_source/MergeWithGitSCMExtension.java<br/>
 src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMBuilderTest.java<br/>
<a href="http://jenkins-ci.org/commit/github-branch-source-plugin/3256676fa6a47d342a0442c59555df9eecde0b68" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/github-branch-source-plugin/3256676fa6a47d342a0442c59555df9eecde0b68</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-45771" title="Disable shallow clone when we know a merge will take place" class="issue-link" data-issue-key="JENKINS-45771"><del>JENKINS-45771</del></a> Pick up consolidated MergeWithGitSCMExtension from git plugin</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=308589&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308589">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><b>Required plugin updates:</b></p>
<ul>
	<li>git-client 2.5.0</li>
	<li>git 3.5.0</li>
	<li>At least one of the optional plugin updates</li>
</ul>


<p><b>Optional plugin updates:</b> (update any of these if already have installed, do not need to install a plugin if not installed already)</p>
<ul>
	<li>github-branch-source 2.2.3</li>
	<li>gitea 1.0.1</li>
	<li>bitbucket-branch-source 2.2.3</li>
</ul>

</pre>
</details>

</ul>
</details>

<p><b>Required plugin updates:</b></p>
<ul>
	<li>git-client 2.5.0</li>
	<li>git 3.5.0</li>
	<li>At least one of the optional plugin updates</li>
</ul>


<p><b>Optional plugin updates:</b> (update any of these if already have installed, do not need to install a plugin if not installed already)</p>
<ul>
	<li>github-branch-source 2.2.3</li>
	<li>gitea 1.0.1</li>
	<li>bitbucket-branch-source 2.2.3</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>shiqiyang:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=335507&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-335507">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I found a case where this bug still applies in parallel stages, reopen this one since I think it is the same issue but please let me know if it is not and I can open another one</p>

<p>environment:
jenkins 2.89.3
github-branch-source-plugin 2.32.2</p>

<p>I have a multibranch project setup with 
`discover the pull requests from origin`
`merge pull request with target`
fetch all tags
shallow clone </p>

<p>I also have Basic Branch Build Strategies Plugin installed to stop the build storm when we scan repo but I don't think it is related</p>

<p>my pipeline file has 2 parallel stages and one of the stages has its own node {} section to specify the builder label and workspace path.
it looks like this (this is almost the same example of <a href="https://jenkins.io/blog/2017/09/25/declarative-1" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/blog/2017/09/25/declarative-1</a>, only difference here is that I have customWorkspace specified on the nodes to reuse the local code repo in the hope to save some time on git clone, our git repo is huge):</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
     agent none
    stages {
        stage(<span class="code-quote">'Run Tests'</span>) {
            parallel {
                stage(<span class="code-quote">'Test 1'</span>) {
                    agent {
                        label <span class="code-quote">"windows"</span>
                        customWorkspace <span class="code-quote">"${env.HOME}/TEST1"</span>
                    }
                    steps {
                        bat <span class="code-quote">"run-tests.bat"</span>
                    }
                    post {
                        always {
                            junit <span class="code-quote">"**/TEST-*.xml"</span>
                        }
                    }
                }
                stage(<span class="code-quote">'Test On Linux'</span>) {
                    agent {
                        label <span class="code-quote">"linux"</span>
                        customWorkspace <span class="code-quote">"${env.HOME}/TEST2"</span>
                    }
                    steps {
                        sh <span class="code-quote">"run-tests.sh"</span>
                    }
                    post {
                        always {
                            junit <span class="code-quote">"**/TEST-*.xml"</span>
                        }
                    }
                }
            }
        }
    }
}
</pre>
</div></div>


<p>the error I am seeing is that when the pipeline enters the parallel stages, especially after the workspace switch, it will fail to merge if the PR is couple of commits behind master because the head of master is checked out without git history. ( I logged in to the builder after the failure and manually verified that in the workspace)</p>

<p>it seems that removing two ways to resolve this for now:
1 remove the the customWorkspace this is not going to work well for us because the lack of self clean on the branches of multibranch (without customworkspace different branch runs will not share workspace in current design of multibranch pipeline)
2. add 'wipe out workspace and clone again in advanced' settings. (this will increase our build time tremendously because our github repo is big, and unnecessary)</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">&gt; /usr/local/bin/git rev-parse --is-inside-work-tree # timeout=10

Fetching changes from the remote Git repository

 &gt; /usr/local/bin/git config remote.origin.url !INFO_EXTRACTED! # timeout=10

Fetching upstream changes from !INFO_EXTRACTED!

 &gt; /usr/local/bin/git --version # timeout=10

using GIT_ASKPASS to set credentials JENKINS_GH

 &gt; /usr/local/bin/git fetch --tags --progress !INFO_EXTRACTED! +refs/pull/!INFO_EXTRACTED!/head:refs/remotes/origin/PR-!INFO_EXTRACTED! +refs/heads/master:refs/remotes/origin/master # timeout=60

Merging remotes/origin/master commit !INFO_EXTRACTED! into PR head commit !INFO_EXTRACTED!

 &gt; /usr/local/bin/git config core.sparsecheckout # timeout=10

 &gt; /usr/local/bin/git checkout -f !INFO_EXTRACTED!

 &gt; /usr/local/bin/git merge !INFO_EXTRACTED! # timeout=10

 &gt; /usr/local/bin/git config core.sparsecheckout # timeout=10

 &gt; /usr/local/bin/git checkout -f !INFO_EXTRACTED!

Command <span class="code-quote">"/usr/local/bin/git merge !INFO_EXTRACTED!"</span> returned status code 128:

stdout: 

stderr: fatal: refusing to merge unrelated histories
</pre>
</div></div>

<p>FYI I remove the sensitive date from the output and replaced with ('!INFO_EXTRACTED')</p>
</pre>
</details>

</ul>
</details>

<p>I found a case where this bug still applies in parallel stages, reopen this one since I think it is the same issue but please let me know if it is not and I can open another one</p>

<p>environment:<br/>
jenkins 2.89.3<br/>
github-branch-source-plugin 2.32.2</p>

<p>I have a multibranch project setup with <br/>
`discover the pull requests from origin`<br/>
`merge pull request with target`<br/>
fetch all tags<br/>
shallow clone </p>

<p>I also have Basic Branch Build Strategies Plugin installed to stop the build storm when we scan repo but I don't think it is related</p>

<p>my pipeline file has 2 parallel stages and one of the stages has its own node {} section to specify the builder label and workspace path.<br/>
it looks like this (this is almost the same example of <a href="https://jenkins.io/blog/2017/09/25/declarative-1" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/blog/2017/09/25/declarative-1</a>, only difference here is that I have customWorkspace specified on the nodes to reuse the local code repo in the hope to save some time on git clone, our git repo is huge):</p>


<pre>
pipeline {
     agent none
    stages {
stage(<span class="code-quote">'Run Tests'</span>) {
    parallel {
stage(<span class="code-quote">'Test 1'</span>) {
    agent {
label <span class="code-quote">"windows"</span>
customWorkspace <span class="code-quote">"${env.HOME}/TEST1"</span>
    }
    steps {
bat <span class="code-quote">"run-tests.bat"</span>
    }
    post {
always {
    junit <span class="code-quote">"**/TEST-*.xml"</span>
}
    }
}
stage(<span class="code-quote">'Test On Linux'</span>) {
    agent {
label <span class="code-quote">"linux"</span>
customWorkspace <span class="code-quote">"${env.HOME}/TEST2"</span>
    }
    steps {
sh <span class="code-quote">"run-tests.sh"</span>
    }
    post {
always {
    junit <span class="code-quote">"**/TEST-*.xml"</span>
}
    }
}
    }
}
    }
}
</pre>


<p>the error I am seeing is that when the pipeline enters the parallel stages, especially after the workspace switch, it will fail to merge if the PR is couple of commits behind master because the head of master is checked out without git history. ( I logged in to the builder after the failure and manually verified that in the workspace)</p>

<p>it seems that removing two ways to resolve this for now:<br/>
1 remove the the customWorkspace this is not going to work well for us because the lack of self clean on the branches of multibranch (without customworkspace different branch runs will not share workspace in current design of multibranch pipeline)<br/>
2. add 'wipe out workspace and clone again in advanced' settings. (this will increase our build time tremendously because our github repo is big, and unnecessary)</p>


<pre>
> /usr/local/bin/git rev-parse --is-inside-work-tree # timeout=10

Fetching changes from the remote Git repository

 > /usr/local/bin/git config remote.origin.url !INFO_EXTRACTED! # timeout=10

Fetching upstream changes from !INFO_EXTRACTED!

 > /usr/local/bin/git --version # timeout=10

using GIT_ASKPASS to set credentials JENKINS_GH

 > /usr/local/bin/git fetch --tags --progress !INFO_EXTRACTED! +refs/pull/!INFO_EXTRACTED!/head:refs/remotes/origin/PR-!INFO_EXTRACTED! +refs/heads/master:refs/remotes/origin/master # timeout=60

Merging remotes/origin/master commit !INFO_EXTRACTED! into PR head commit !INFO_EXTRACTED!

 > /usr/local/bin/git config core.sparsecheckout # timeout=10

 > /usr/local/bin/git checkout -f !INFO_EXTRACTED!

 > /usr/local/bin/git merge !INFO_EXTRACTED! # timeout=10

 > /usr/local/bin/git config core.sparsecheckout # timeout=10

 > /usr/local/bin/git checkout -f !INFO_EXTRACTED!

Command <span class="code-quote">"/usr/local/bin/git merge !INFO_EXTRACTED!"</span> returned status code 128:

stdout: 

stderr: fatal: refusing to merge unrelated histories
</pre>

<p>FYI I remove the sensitive date from the output and replaced with ('!INFO_EXTRACTED')</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=352608&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-352608">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Removing myself as assignee. My current work assignments do not provide sufficient bandwidth to review these issues and in the majority of cases I am only assigned by virtue of being the default assignee. For the credentials-api and scm-api related plugins I have permission to allocate time reviewing changes to these APIs themselves to ensure these APIs remain cohesive, but that can be handled through PR reviews rather than assigning issues in JIRA</p>
</pre>
</details>

</ul>
</details>

<p>Removing myself as assignee. My current work assignments do not provide sufficient bandwidth to review these issues and in the majority of cases I am only assigned by virtue of being the default assignee. For the credentials-api and scm-api related plugins I have permission to allocate time reviewing changes to these APIs themselves to ensure these APIs remain cohesive, but that can be handled through PR reviews rather than assigning issues in JIRA</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nfalco.png" title="nfalco's avatar" /> nfalco:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=450981&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-450981">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I see all PR has been merged, could be this issue closed?</p>
</pre>
</details>

</ul>
</details>

<p>I see all PR has been merged, could be this issue closed?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45771?focusedId=450984&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-450984">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've done no investigation to see if the two repository use case described by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shiqiyang" class="user-hover" rel="shiqiyang">shiqiyang</a> is an issue.  I don't plan to investigate that use case because I don't plan to take any action on that use case, even if it is an issue.  I think that users who want to merge before release should use Pipeline and perform the merge from an sh step rather than relying on the git plugin to apply its custom logic for a merge.  I don't plan to remove the feature from the git plugin, but I don't plan to invest significant effort to improve it either.</p>
</pre>
</details>

</ul>
</details>

<p>I've done no investigation to see if the two repository use case described by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shiqiyang" class="user-hover" rel="shiqiyang">shiqiyang</a> is an issue.  I don't plan to investigate that use case because I don't plan to take any action on that use case, even if it is an issue.  I think that users who want to merge before release should use Pipeline and perform the merge from an sh step rather than relying on the git plugin to apply its custom logic for a merge.  I don't plan to remove the feature from the git plugin, but I don't plan to invest significant effort to improve it either.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-44598%22">JENKINS-44598</a>]</i>
